components/automate-deployment/pkg/manifest/provider.go

```diff
@@ -174,3 +188,258 @@ func (provider *LocalHartManifestProvider) overlayLocalHarts(baseManifest *A2) (
+	allVersions = allVersions[currentVersionIndex:]
+
+	//get the minor/patch versions are available or not.
+	//check the current version is timestamp or semantic version
+	currentMajor, isSemVersion := IsSemVersionFmt(version)
+
+	if !isSemVersion {
+		//check for any minor
+		index, timeStampVersion := findLatestTimeStampVersion(allVersions)
+		if index != 0 {
+			return true, false, timeStampVersion, nil
+		}
+		_, nextMajorVersion := findNextMajorVersionForTimeStamp(allVersions)
+		return false, true, nextMajorVersion, nil
+	} else {
+		index, patchVersion := findPatchVersionForSemantic(currentMajor, allVersions)
+		if index != 0 {
+			return true, false, patchVersion, nil
+		}
+		_, nextMajorVersion := findNextMajorVersionForSemantic(currentMajor, allVersions)
+		return false, true, nextMajorVersion, nil
+	}
+}
+
+//isSemVersionFmt checks the provided version is in semantic version format, if yes, will return the major version
+func IsSemVersionFmt(version string) (string, bool) {
+	splitStrings := strings.Split(version, ".")
+	if len(splitStrings) > 1 {
+		return splitStrings[0], true
+	}
+	return "", false
+}
+
+func findLatestTimeStampVersion(list []string) (index int, version string) {
+	for index, item := range list {
+		if _, isSem := IsSemVersionFmt(item); isSem {
+			return index - 1, list[index-1]
+		} else if index == len(list)-1 { //reached to end
+			return index, list[index]
+		}
+	}
+	//ideally, this should not execute
+	return 0, list[0]
+}
+
+func findNextMajorVersionForTimeStamp(list []string) (int, string) {
+	majorVersion := ""
+	for index, item := range list {
+		if major, isSem := IsSemVersionFmt(item); isSem {
+			if majorVersion == "" {
+				majorVersion = major
+			} else if majorVersion != major {
+				return index - 1, list[index-1]
+			}
+			if index == len(list)-1 && majorVersion != "" {
+				return index, list[index]
+			}
+		}
+	}
+	return 0, ""
+}
+
+func findPatchVersionForSemantic(currentMajor string, list []string) (int, string) {
+	for index, item := range list {
+		if major, _ := IsSemVersionFmt(item); currentMajor != major {
+			return index - 1, list[index-1]
+		} else if index == len(list)-1 { //reached to end
+			return index, list[index]
+		}
+	}
+	//ideally, this should not execute
+	return 0, list[0]
+}
+
+func findNextMajorVersionForSemantic(currentMajor string, list []string) (int, string) {
+	nextMajorVersion := ""
+	for index, item := range list {
+		if major, _ := IsSemVersionFmt(item); currentMajor != major {
+			if nextMajorVersion == "" {
+				nextMajorVersion = major
+			} else if nextMajorVersion != major {
+				return index - 1, list[index-1]
+			}
+			if nextMajorVersion != "" && index == len(list)-1 {
+				return index, list[index]
+			}
+		}
+	}
+	return 0, ""
+}
+
+//GetAllVersions gives the list of all released versions of given channel in ascending order of version numbers
+func GetAllVersions(ctx context.Context, channel string, optionalURL ...string) ([]string, error) {
```

Comment:
should this live in http.go along with the other manifest urls,  instead of here?

Reply:
No @rickmarry, that should go to http.go only when we are implementing this at the client level. But here we are implementing at the manifest provider level, and not at the client level, so this should be here only.